src/shogun/lib/any.h

```diff
@@ -1329,34 +1349,34 @@ namespace shogun
 			if constexpr (std::is_base_of_v<SGObject, SharedType>)
 			{
 				Any::register_caster<T, std::shared_ptr<SGObject>>(
-						[](T value) { return std::dynamic_pointer_cast<SGObject>(value); });
+						[](const T& value) { return std::dynamic_pointer_cast<SGObject>(value); });
 				if constexpr (!std::is_same_v<std::nullptr_t, base_type<SharedType>>
 						&& !std::is_same_v<SharedType, base_type<SharedType>>)
-					Any::register_caster<T, std::shared_ptr<base_type<SharedType>>>([](T value) {
+					Any::register_caster<T, std::shared_ptr<base_type<SharedType>>>([](const T& value) {
 						return std::dynamic_pointer_cast<base_type<SharedType>>(value);
 					});
 			}
 		}
 		if constexpr (std::is_arithmetic_v<T>)
 		{
 			Any::register_caster<T, float32_t>(
-			    [](T value) { return utils::safe_convert<float32_t>(value); });
+			    [](const T& value) { return utils::safe_convert<float32_t>(value); });
 			Any::register_caster<T, float64_t>(
-			    [](T value) { return utils::safe_convert<float64_t>(value); });
+			    [](const T& value) { return utils::safe_convert<float64_t>(value); });
 			Any::register_caster<T, int32_t>(
-			    [](T value) { return utils::safe_convert<int32_t>(value); });
+			    [](const T& value) { return utils::safe_convert<int32_t>(value); });
 			Any::register_caster<T, int64_t>(
-			    [](T value) { return utils::safe_convert<int64_t>(value); });
+			    [](const T& value) { return utils::safe_convert<int64_t>(value); });
 		}
 		if constexpr (traits::has_std_to_string<T>::value)
 		{
 			Any::register_caster<T, std::string>(
-			    [](T value) { return std::to_string(value); });
+			    [](const T& value) { return std::to_string(value); });
 		}
 		if constexpr (traits::has_to_string<T>::value)
 		{
 			Any::register_caster<T, std::string>(
-			    [](T value) { return value.to_string(); });
+			    [](const T& value) { return value.to_string(); });
```

Comment:
made const ref because you cannot copy std::atomic, but it also doesn't make sense to copy all these values into the casting function